printing/printer.ml

```diff
@@ -342,15 +349,55 @@ let pr_ne_context_of header env sigma =
     List.is_empty (Environ.named_context env)  then (mt ())
   else let penv = pr_context_unlimited env sigma in (header ++ penv ++ fnl ())
 
+(* Heuristic for horizontalizing hypothesis:
+   Detecting variable which type is a simple id or of the form (t x y ...)
+   where t is a product or only sorts (typically [Type -> Type -> ...]
+   and not [nat -> nat -> ...] ).
+   + Special case for non-Prop dependent terms. *)
+let rec should_compact env sigma typ =
+  get_compact_context() &&
+    match kind_of_term typ with
+    | Rel _ | Var _ | Sort _ | Const _ | Ind _ -> true
+    | App (c,args) ->
+       let _,type_of_c = Typing.type_of env sigma (EConstr.of_constr c) in
+       let _,type_of_typ = Typing.type_of env sigma (EConstr.of_constr typ) in
+       not (is_Prop (EConstr.to_constr sigma type_of_typ))
+       && (* These two more tests detect rare cases of non-Prop-sorted
+             dependent hypothesis:  *)
+       let lnamedtyp , _ = EConstr.decompose_prod sigma type_of_c in
+       (* c has a non dependent type *)
```

Comment:
OK let us say it another way: I want to distinguish what the user considers a "variable" from what he considers a "hypothesis". The first category can be compacted (or even hidden as suggested by @jonleivent), the second category is nicer with one hyp per line.

The test is naive but will only consider some  non-dependent types as hypothesis I think, which means that compaction will happen less often than it could. I want to avoid the opposite (compacting something that the user consider as an "hypothesis"). It proved ok when I used it for some months but it may be a little subjective.

If someone has a better criterion to make this distinction I am open to suggestions.